src/isomorphism/vf2.rs

```diff
@@ -41,6 +38,86 @@ use crate::iterators::NodeMap;
 
 type StablePyGraph<Ty> = StableGraph<PyObject, PyObject, Ty>;
 
+/// Returns `true` if we can map every element of `xs` to a unique
+/// element of `ys` while using `matcher` func to compare two elements.
+fn is_subset<T: Copy, F>(xs: &[T], ys: &[T], matcher: F) -> PyResult<bool>
+where
+    F: Fn(T, T) -> PyResult<bool>,
+{
+    let mut valid = vec![true; ys.len()];
+    for &a in xs {
+        let mut found = false;
+        for (&b, free) in ys.iter().zip(valid.iter_mut()) {
+            if *free && matcher(a, b)? {
+                found = true;
+                *free = false;
+                break;
+            }
+        }
+
+        if !found {
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
+
+#[inline]
+fn sorted<N: std::cmp::PartialOrd>(x: &mut (N, N)) {
+    let (a, b) = x;
+    if b < a {
+        std::mem::swap(a, b)
+    }
+}
+
+/// Returns the adjacency matrix of a graph as a dictionary
+/// with `(i, j)` entry equal to number of edges from node `i` to node `j`.
+fn adjacency_matrix<Ty: EdgeType>(
+    graph: &StablePyGraph<Ty>,
+) -> HashMap<(NodeIndex, NodeIndex), usize> {
+    let mut matrix = HashMap::with_capacity(graph.edge_count());
```

Comment:
Won't this potentially overallocate in the case of parallel edges?

Reply:
Hmm, yeah, it might over allocate but probably by not that much. It's quite difficult to obtain a more tight bound in the number of non-parallel edges. Are you suggesting to not pre-allocate at all?

Reply:
Yeah, that's what I was getting at. But, thinking about it some more I agree it's probably not a bid deal we basically overallocate 3 usizes for each parallel edge in the graph which will probably hardly be noticeable. Especially compared to the extra allocation overhead for not pre-allocating.